File: old -backend/src/middleware/rate.limiter.js
/**
 * RATE LIMITER MIDDLEWARE
 * Prevents brute-force attacks on login and API abuse.
 *
 * Uses in-memory store (fine for single-process Render.com free tier).
 * No external Redis needed.
 */

// Simple in-memory rate limit store
// Map<key, { count, resetAt }>
const store = new Map();

/**
 * createLimiter(options)
 * @param {number} windowMs   - Time window in ms
 * @param {number} max        - Max requests per window
 * @param {string} message    - Error message
 * @param {function} keyFn    - Function to extract key from req (default: IP)
 */
function createLimiter({ windowMs = 60_000, max = 60, message = 'Too many requests', keyFn } = {}) {
  return (req, res, next) => {
    const key = keyFn ? keyFn(req) : req.ip;
    const now = Date.now();

    let entry = store.get(key);
    if (!entry || now > entry.resetAt) {
      entry = { count: 0, resetAt: now + windowMs };
      store.set(key, entry);
    }

    entry.count++;

    res.setHeader('X-RateLimit-Limit', max);
    res.setHeader('X-RateLimit-Remaining', Math.max(0, max - entry.count));
    res.setHeader('X-RateLimit-Reset', Math.ceil(entry.resetAt / 1000));

    if (entry.count > max) {
      console.warn(`🚫 RATE LIMIT HIT | Key: ${key} | Count: ${entry.count}`);
      return res.status(429).json({
        error: message,
        retry_after_seconds: Math.ceil((entry.resetAt - now) / 1000),
      });
    }

    next();
  };
}

// Clean up expired entries every 5 minutes (prevent memory leak)
setInterval(() => {
  const now = Date.now();
  for (const [key, val] of store.entries()) {
    if (now > val.resetAt) store.delete(key);
  }
}, 5 * 60 * 1000);

// ── Pre-configured limiters ───────────────────────────────────────────────

// Login: 5 attempts per 15 minutes per IP
const loginLimiter = createLimiter({
  windowMs: 15 * 60_000,
  max: 5,
  message: 'Too many login attempts. Please try again in 15 minutes.',
});

// General API: 120 req/min per authenticated user (by profile ID)
const apiLimiter = createLimiter({
  windowMs: 60_000,
  max: 120,
  message: 'API rate limit exceeded. Please slow down.',
  keyFn: (req) => req.profileId || req.ip,
});

// Strict limiter for sensitive ops (approve, generate cert): 20/min
const sensitiveLimiter = createLimiter({
  windowMs: 60_000,
  max: 20,
  message: 'Too many sensitive operations. Please slow down.',
  keyFn: (req) => req.profileId || req.ip,
});

module.exports = { createLimiter, loginLimiter, apiLimiter, sensitiveLimiter };

File: dart_backend/lib/middleware/rate_limiter.dart
// lib/middleware/rate_limiter.dart
// In-memory rate limiter — port of rate.limiter.js
// No external Redis needed for single-process Render.com free tier.

import 'dart:async';
import 'dart:convert';
import 'package:shelf/shelf.dart';
import '../models/user_session.dart';

class _Entry {
  int count;
  DateTime resetAt;
  _Entry({required this.count, required this.resetAt});
}

class RateLimiter {
  final Duration window;
  final int max;
  final String message;
  final String Function(Request req)? keyFn;

  final Map<String, _Entry> _store = {};
  RateLimiter({
    required this.window,
    required this.max,
    required this.message,
    this.keyFn,
  }) {
    // Clean up expired entries every 5 minutes
    Timer.periodic(const Duration(minutes: 5), (_) {
      final now = DateTime.now();
      _store.removeWhere((_, v) => now.isAfter(v.resetAt));
    });
  }

  Middleware get middleware {
    return (Handler inner) {
      return (Request request) async {
        final session = request.context['session'] as UserSession?;
        final key = keyFn != null
            ? keyFn!(request)
            : (session?.profileId ?? _getIp(request));

        final now = DateTime.now();
        _Entry? entry = _store[key];
        if (entry == null || now.isAfter(entry.resetAt)) {
          entry = _Entry(count: 0, resetAt: now.add(window));
          _store[key] = entry;
        }

        entry.count++;

        final remaining = (max - entry.count).clamp(0, max);
        final resetEpoch = entry.resetAt.millisecondsSinceEpoch ~/ 1000;

        final headers = {
          'X-RateLimit-Limit': max.toString(),
          'X-RateLimit-Remaining': remaining.toString(),
          'X-RateLimit-Reset': resetEpoch.toString(),
        };

        if (entry.count > max) {
          print('🚫 RATE LIMIT HIT | Key: $key | Count: ${entry.count}');
          final retryAfter = entry.resetAt.difference(now).inSeconds;
          return Response(429,
              body: jsonEncode({'error': message, 'retry_after_seconds': retryAfter}),
              headers: {...headers, 'Content-Type': 'application/json'});
        }

        // Add headers to the response from inner handler
        final response = await inner(request);
        return response.change(headers: {...response.headersAll, ...headers.map((k, v) => MapEntry(k, [v]))});
      };
    };
  }

  String _getIp(Request req) =>
      req.headers['x-forwarded-for']?.split(',').first.trim() ??
      req.headers['x-real-ip'] ??
      'unknown';
}

// ── Pre-configured limiters (mirrors rate.limiter.js) ────────────────────────

/// Login: 5 attempts per 15 minutes per IP
final loginLimiter = RateLimiter(
  window: const Duration(minutes: 15),
  max: 5,
  message: 'Too many login attempts. Please try again in 15 minutes.',
);

/// General API: 120 req/min per authenticated user
final apiLimiter = RateLimiter(
  window: const Duration(minutes: 1),
  max: 120,
  message: 'API rate limit exceeded. Please slow down.',
  keyFn: (req) {
    final session = req.context['session'] as UserSession?;
    return session?.profileId ?? 'anon';
  },
);

/// Sensitive ops: 20/min (approve, generate cert)
final sensitiveLimiter = RateLimiter(
  window: const Duration(minutes: 1),
  max: 20,
  message: 'Too many sensitive operations. Please slow down.',
  keyFn: (req) {
    final session = req.context['session'] as UserSession?;
    return session?.profileId ?? 'anon';
  },
);
